File: java/src/main/java/ai/guard8/shield/TOTP.java
package ai.guard8.shield;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.Base64;

/**
 * TOTP - Time-based One-Time Password (RFC 6238)
 */
public class TOTP {
    public static final int DEFAULT_DIGITS = 6;
    public static final int DEFAULT_INTERVAL = 30;
    public static final int DEFAULT_SECRET_SIZE = 20;

    private final byte[] secret;
    private final int digits;
    private final long interval;

    public TOTP(byte[] secret) {
        this(secret, DEFAULT_DIGITS, DEFAULT_INTERVAL);
    }

    public TOTP(byte[] secret, int digits, int interval) {
        this.secret = Arrays.copyOf(secret, secret.length);
        this.digits = digits > 0 ? digits : DEFAULT_DIGITS;
        this.interval = interval > 0 ? interval : DEFAULT_INTERVAL;
    }

    public static byte[] generateSecret() {
        return Shield.randomBytes(DEFAULT_SECRET_SIZE);
    }

    public String generate() {
        return generate(System.currentTimeMillis() / 1000);
    }

    public String generate(long timestamp) {
        if (timestamp == 0) {
            timestamp = System.currentTimeMillis() / 1000;
        }
        long counter = timestamp / interval;
        return generateHOTP(counter);
    }

    public boolean verify(String code) {
        return verify(code, 0, 1);
    }

    public boolean verify(String code, long timestamp, int window) {
        if (timestamp == 0) {
            timestamp = System.currentTimeMillis() / 1000;
        }
        if (window <= 0) {
            window = 1;
        }

        for (int i = 0; i <= window; i++) {
            // Check current and past
            if (generate(timestamp - i * interval).equals(code)) {
                return true;
            }
            // Check future (except for i=0)
            if (i > 0 && generate(timestamp + i * interval).equals(code)) {
                return true;
            }
        }
        return false;
    }

    private String generateHOTP(long counter) {
        byte[] counterBytes = ByteBuffer.allocate(8).putLong(counter).array();

        try {
            Mac mac = Mac.getInstance("HmacSHA1");
            mac.init(new SecretKeySpec(secret, "HmacSHA1"));
            byte[] hash = mac.doFinal(counterBytes);

            int offset = hash[19] & 0x0f;
            int code = ((hash[offset] & 0x7f) << 24) |
                       ((hash[offset + 1] & 0xff) << 16) |
                       ((hash[offset + 2] & 0xff) << 8) |
                       (hash[offset + 3] & 0xff);

            int modulo = 1;
            for (int i = 0; i < digits; i++) {
                modulo *= 10;
            }

            return String.format("%0" + digits + "d", code % modulo);
        } catch (Exception e) {
            throw new RuntimeException("HMAC-SHA1 not available", e);
        }
    }

    public String toBase32() {
        return Base32.encode(secret);
    }

    public static TOTP fromBase32(String encoded) {
        return new TOTP(Base32.decode(encoded));
    }

    public String getProvisioningUri(String account, String issuer) {
        String secretB32 = toBase32();
        return String.format(
            "otpauth://totp/%s:%s?secret=%s&issuer=%s&algorithm=SHA1&digits=%d&period=%d",
            issuer, account, secretB32, issuer, digits, interval
        );
    }

    public byte[] getSecret() {
        return Arrays.copyOf(secret, secret.length);
    }

    public void wipe() {
        Arrays.fill(secret, (byte) 0);
    }

    // Simple Base32 encoder/decoder
    private static class Base32 {
        private static final String ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

        public static String encode(byte[] data) {
            StringBuilder result = new StringBuilder();
            int buffer = 0;
            int bufferLength = 0;

            for (byte b : data) {
                buffer = (buffer << 8) | (b & 0xff);
                bufferLength += 8;
                while (bufferLength >= 5) {
                    bufferLength -= 5;
                    result.append(ALPHABET.charAt((buffer >> bufferLength) & 0x1f));
                }
            }
            if (bufferLength > 0) {
                result.append(ALPHABET.charAt((buffer << (5 - bufferLength)) & 0x1f));
            }
            return result.toString();
        }

        public static byte[] decode(String encoded) {
            encoded = encoded.toUpperCase().replaceAll("=", "");
            byte[] result = new byte[encoded.length() * 5 / 8];
            int buffer = 0;
            int bufferLength = 0;
            int index = 0;

            for (char c : encoded.toCharArray()) {
                int val = ALPHABET.indexOf(c);
                if (val < 0) continue;
                buffer = (buffer << 5) | val;
                bufferLength += 5;
                if (bufferLength >= 8) {
                    bufferLength -= 8;
                    result[index++] = (byte) ((buffer >> bufferLength) & 0xff);
                }
            }
            return Arrays.copyOf(result, index);
        }
    }
}

File: kotlin/src/main/kotlin/ai/guard8/shield/TOTP.kt
package ai.guard8.shield

import java.nio.ByteBuffer
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * TOTP - Time-based One-Time Password (RFC 6238)
 */
class TOTP(
    private val secret: ByteArray,
    private val digits: Int = DEFAULT_DIGITS,
    private val interval: Long = DEFAULT_INTERVAL
) : AutoCloseable {

    companion object {
        const val DEFAULT_DIGITS = 6
        const val DEFAULT_INTERVAL = 30L
        const val DEFAULT_SECRET_SIZE = 20

        fun generateSecret(): ByteArray = Shield.randomBytes(DEFAULT_SECRET_SIZE)

        fun fromBase32(encoded: String): TOTP = TOTP(Base32.decode(encoded))
    }

    fun generate(timestamp: Long = System.currentTimeMillis() / 1000): String {
        val counter = timestamp / interval
        return generateHOTP(counter)
    }

    fun verify(code: String, timestamp: Long = System.currentTimeMillis() / 1000, window: Int = 1): Boolean {
        for (i in 0..window) {
            if (generate(timestamp - i * interval) == code) return true
            if (i > 0 && generate(timestamp + i * interval) == code) return true
        }
        return false
    }

    private fun generateHOTP(counter: Long): String {
        val counterBytes = ByteBuffer.allocate(8).putLong(counter).array()

        val mac = Mac.getInstance("HmacSHA1")
        mac.init(SecretKeySpec(secret, "HmacSHA1"))
        val hash = mac.doFinal(counterBytes)

        val offset = hash[19].toInt() and 0x0f
        val code = ((hash[offset].toInt() and 0x7f) shl 24) or
                   ((hash[offset + 1].toInt() and 0xff) shl 16) or
                   ((hash[offset + 2].toInt() and 0xff) shl 8) or
                   (hash[offset + 3].toInt() and 0xff)

        var modulo = 1
        repeat(digits) { modulo *= 10 }

        return (code % modulo).toString().padStart(digits, '0')
    }

    fun toBase32(): String = Base32.encode(secret)

    fun provisioningUri(account: String, issuer: String): String {
        val secretB32 = toBase32()
        return "otpauth://totp/$issuer:$account?secret=$secretB32&issuer=$issuer&algorithm=SHA1&digits=$digits&period=$interval"
    }

    fun getSecret(): ByteArray = secret.copyOf()

    override fun close() {
        Shield.secureWipe(secret)
    }
}

/**
 * Recovery codes for backup authentication.
 */
class RecoveryCodes(count: Int = 10) {
    private val codes = mutableSetOf<String>()

    init {
        repeat(count) {
            codes.add(generateCode())
        }
    }

    private fun generateCode(): String {
        val bytes = Shield.randomBytes(4)
        val part1 = ((bytes[0].toInt() and 0xff) shl 8) or (bytes[1].toInt() and 0xff)
        val part2 = ((bytes[2].toInt() and 0xff) shl 8) or (bytes[3].toInt() and 0xff)
        return String.format("%04X-%04X", part1, part2)
    }

    fun verify(code: String): Boolean {
        val normalized = code.uppercase().replace(" ", "")
        val formatted = if (normalized.length == 8) "${normalized.take(4)}-${normalized.takeLast(4)}" else normalized

        return if (codes.contains(formatted)) {
            codes.remove(formatted)
            true
        } else {
            false
        }
    }

    val allCodes: List<String> get() = codes.toList()
    val remaining: Int get() = codes.size
}

// Simple Base32 encoder/decoder
private object Base32 {
    private const val ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"

    fun encode(data: ByteArray): String {
        val result = StringBuilder()
        var buffer = 0
        var bufferLength = 0

        for (byte in data) {
            buffer = (buffer shl 8) or (byte.toInt() and 0xff)
            bufferLength += 8
            while (bufferLength >= 5) {
                bufferLength -= 5
                result.append(ALPHABET[(buffer shr bufferLength) and 0x1f])
            }
        }
        if (bufferLength > 0) {
            result.append(ALPHABET[(buffer shl (5 - bufferLength)) and 0x1f])
        }
        return result.toString()
    }

    fun decode(encoded: String): ByteArray {
        val clean = encoded.uppercase().replace("=", "")
        val result = ByteArray(clean.length * 5 / 8)
        var buffer = 0
        var bufferLength = 0
        var index = 0

        for (char in clean) {
            val value = ALPHABET.indexOf(char)
            if (value < 0) continue
            buffer = (buffer shl 5) or value
            bufferLength += 5
            if (bufferLength >= 8) {
                bufferLength -= 8
                result[index++] = ((buffer shr bufferLength) and 0xff).toByte()
            }
        }
        return result.copyOf(index)
    }
}
